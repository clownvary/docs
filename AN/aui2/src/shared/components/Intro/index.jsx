import React from 'react';
import ReactDOM from 'react-dom';
import UIComponent from 'shared/components/UIComponent';
import Button from 'react-base-ui/lib/components/Button';

import './index.less';
import Step from './step';

export default class Intro extends UIComponent {
  constructor(props) {
    super(props);
    this.setState = { update: false };
  }

  render() {
    // please keep 'steps, hide, ready, onClickCallback' here, otherwise attrs will have this prop
    // it's a unknown prop for DOM, and will has a warning in browser
    // eslint-disable-next-line no-unused-vars
    const { steps, hide, ready, onClickCallback, ...attrs } = this.props;

    return React.DOM.div(attrs);
  }

  componentDidMount() {
    this.intro = document.createElement('div');
    document.body.appendChild(this.intro);
    this._render();
    window.addEventListener('resize', this.onWindowResize);
  }

  componentDidUpdate() {
    const { hide, ready } = this.props;
    /* istanbul ignore else */
    if (!hide && ready) {
      // hide body's scroll-bar
      document.body.style.overflow = 'hidden';
    }

    this._render();
  }

  componentWillUnmount() {
    this.unmount();
  }

  onWindowResize = () => {
    this._render();
  }

  onClickCallback = () => {
    // enable body's scroll-bar
    document.body.style.overflow = 'auto';

    const { onClickCallback } = this.props;
    /* istanbul ignore else */
    if (onClickCallback) {
      onClickCallback();
    }
  }

  _render = () => {
    const { hide, ready, steps } = this.props;
    const update = this.update;
    this.update = !this.update;

    ReactDOM.render(<div className={`intro ${(!hide && ready) ? '' : 'u-hidden'}`}>
      <div className="intro-overlay" />
      {
        steps.map((step, i) => {
          const target = document.querySelector(step.element);
          if (!target) {
            return null;
          }
          return <Step step={{ ...step, target }} key={i} update={update} />;
        })
      }{this.renderControl()}</div>, this.intro);
  }

  unmount = () => {
    ReactDOM.unmountComponentAtNode(this.intro);
    document.body.removeChild(this.intro);
    window.removeEventListener('resize', this.onWindowResize);
  }

  renderControl = () => {
    const { children, closeButtonText } = this.props;
    if (children) {
      return <div className="intro-control-btn" onClick={this.onClickCallback}>{children}</div>;
    }
    return (<Button
      className="intro-control-btn" type="strong"
      onClick={this.onClickCallback}
    >{closeButtonText || 'Got It'}</Button>);
  }
}

Intro.defaultProps = {
  steps: [],
  onClickCallback() {
    this.unmount();
  }
};
