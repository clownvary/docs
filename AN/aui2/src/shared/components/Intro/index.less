@step-element-z-index: 1000;
@show-element-z-index: @step-element-z-index + 2;
@step-mask-z-index: @step-element-z-index + 4;
@interval: 35px;

.intro {
  &-overlay {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 999;
    background-color: #000;
    opacity: 0.8;
  }

  &-showElement {
    position: relative;
    z-index: @show-element-z-index;
  }

  &-step-wrapper {
    position: absolute;
  }

  &-step,
  &-step-mask {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }

  &-step {
    z-index: @step-element-z-index;
    box-shadow: 0 2px 15px rgba(0, 0, 0, 0.4);
    box-sizing: border-box;
    background-color: #fff;
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 5px;
  }

  &-step-mask {
    z-index: @step-mask-z-index;
    background: transparent;
  }

  &-tooltip {
    position: absolute;
    display: flex;
    font-style: oblique;
    white-space: nowrap;
    color: #d8d5d5;

    .intro-arrow {
      position: absolute;
    }

    &-right {
      top: 50px;
      left: 100%;
      margin-left: @interval;

      .intro-arrow {
        top: -30px;
        left: 0;
        transform: rotate(90deg) scaleX(-1);
      }
    }

    &-left {
      top: 50px;
      right: 100%;
      margin-right: @interval;

      .intro-arrow {
        top: -30px;
        right: 0;
        transform: rotate(-90deg);
      }
    }

    &-left-bottom,
    &-right-bottom,
    &-middle-bottom {
      top: 100%;
      margin-top: @interval;

      .intro-arrow {
        bottom: 50%;
      }
    }

    &-left-bottom {
      left: 50px;

      .intro-arrow {
        left: -30px;
        transform: rotate(180deg);
      }
    }

    &-right-bottom {
      right: 50px;

      .intro-arrow {
        right: -30px;
        transform: rotate(180deg) scaleX(-1);
      }
    }

    &-middle-bottom {
      left: 50%;
      transform: translate(-60%, 0);

      .intro-arrow {
        right: -30px;
        transform: rotate(180deg) scaleX(-1);
      }
    }

    &-left-top,
    &-right-top,
    &-middle-top {
      bottom: 100%;
      align-items: flex-end;
      margin-bottom: @interval;

      .intro-arrow {
        top: 50%;
      }
    }

    &-left-top {
      left: 50px;

      .intro-arrow {
        left: -30px;
        transform: scaleX(-1);
      }
    }

    &-right-top {
      right: 50px;

      .intro-arrow {
        right: -30px;
      }
    }

    &-middle-top {
      left: 50%;
      transform: translate(-60%, 0);

      .intro-arrow {
        right: -30px;
      }
    }
  }

  &-display-inline-block {
    display: inline-block;
  }

  &-control-btn {
    position: fixed;
    top: 50%;
    left: 50%;
    z-index: @step-element-z-index;
    transform: translate(-50%, -50%);
  }
}
