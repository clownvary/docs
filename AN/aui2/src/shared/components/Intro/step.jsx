import React from 'react';
import UIComponent from 'shared/components/UIComponent';
import IntroArrow from 'shared/components/IntroArrow';

const timeout = 300;

export default class Step extends UIComponent {
  constructor(props) {
    super(props);
    this.timer = 0;
    this.state = {
      left: 0,
      top: 0,
      width: 0,
      height: 0,
      display: 'none'
    };
  }

  render() {
    const { step } = this.props;

    return (
      <div className="intro-step-wrapper" style={this.state}>
        <div className={`intro-step ${step.className || ''}`}>
          <div className={`intro-tooltip intro-tooltip-${step.position}`}>
            {step.text}
            <IntroArrow />
          </div>
        </div>
        <div className="intro-step-mask" />
      </div>
    );
  }

  static getOffset(elem) {
    if (!elem.getClientRects().length) {
      return { top: 0, left: 0 };
    }

    const rect = elem.getBoundingClientRect();

    // Make sure element is not hidden (display: none)
    /* istanbul ignore else */
    if (rect.width || rect.height) {
      const doc = elem.ownerDocument;
      /* istanbul ignore next */
      const win = doc === doc.window ? doc : doc.nodeType === 9 && doc.defaultView;
      const docElem = doc.documentElement;
      return {
        top: (rect.top + win.pageYOffset) - docElem.clientTop,
        left: (rect.left + win.pageXOffset) - docElem.clientLeft
      };
    }

    return rect;
  }

  setPosition = (target) => {
    const { left: prevLeft, top: prevTop } = Step.getOffset(target);
    this.timer = setTimeout(() => {
      const { left, top } = Step.getOffset(target);
      const padding = 2;

      if (left === prevLeft && top === prevTop) {
        this.setState({
          left: left - padding,
          top: top - padding,
          width: target.offsetWidth + (padding * 2),
          height: target.offsetHeight + (padding * 2),
          display: 'block'
        });
      } else {
        this.setPosition(target);
      }
    }, timeout);
  }

  componentDidMount() {
    const { target } = this.props.step;
    target.className += ' intro-showElement';

    /* istanbul ignore if */
    if (window.getComputedStyle && window.getComputedStyle(target, null).getPropertyValue('display') === 'inline') {
      target.className += ' intro-display-inline-block';
    }
    this.setPosition(target);
  }

  componentWillUpdate() {
    this.setPosition(this.props.step.target);
  }

  componentWillUnmount() {
    const target = this.props.step.target;

    clearTimeout(this.timer);

    target.className = target.className.split(' ').reduce((prevC, currentC) => {
      if (currentC.indexOf('intro-showElement') !== -1 || currentC.indexOf('intro-display-inline-block') !== -1) {
        return prevC;
      }
      return `${prevC} ${currentC}`;
    });
  }
}
