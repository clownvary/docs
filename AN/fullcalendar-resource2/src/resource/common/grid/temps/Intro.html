<td class="fc-week-number" {{{weekNumberStyleAttr}}}></td>
