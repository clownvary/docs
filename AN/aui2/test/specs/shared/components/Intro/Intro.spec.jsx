import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Intro from 'shared/components/Intro';

jest.mock('shared/components/Intro/step', () => 'Step');

describe('shared/components/Intro/index', () => {
  let stepTarget;
  const id1 = 'intro-step-1';

  const props = {
    steps: [
      {
        text: '1. First step...',
        element: `#${id1}`
      },
      {
        text: '2. Second step...(invalid)',
        element: '#invalid-id'
      }
    ],
    hide: true,
    ready: false,
    children: null
  };

  beforeEach(() => {
    stepTarget = document.createElement('div');
    stepTarget.setAttribute('id', id1);
    document.body.appendChild(stepTarget);
  });

  afterEach(() => {
    stepTarget = document.getElementById(id1);
    document.body.removeChild(stepTarget);

    const div = document.querySelector('div');
    div && document.body.removeChild(div);
  });

  it('should render without errors', () => {
    const component = mount(<Intro {...props} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('component should work fine', () => {
    const controlBtnClick = jest.fn();
    const component = mount(<Intro {...props} onClickCallback={controlBtnClick} />);

    expect(document.querySelector('.intro').className.indexOf('u-hidden')).toBeGreaterThanOrEqual(0);

    const newProps = Object.assign({}, props, { hide: false, ready: true, children: 'Intro...' });
    component.setProps(newProps);

    expect(document.querySelector('.intro').className.indexOf('u-hidden')).toBeLessThan(0);

    document.querySelector('.intro-control-btn').click();
    expect(controlBtnClick).toHaveBeenCalled();

    window.dispatchEvent(new Event('resize'))

    component.unmount();
  });

  it('component should unmount correctly if not giving click callback', () => {
    const component = mount(<Intro {...props} />);
    document.querySelector('.intro-control-btn').click();
  });
});
