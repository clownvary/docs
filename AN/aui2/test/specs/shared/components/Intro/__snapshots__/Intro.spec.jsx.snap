// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared/components/Intro/index should render without errors 1`] = `
<Intro
  hide={true}
  onClickCallback={[Function]}
  ready={false}
  steps={
    Array [
      Object {
        "element": "#intro-step-1",
        "text": "1. First step...",
      },
      Object {
        "element": "#invalid-id",
        "text": "2. Second step...(invalid)",
      },
    ]
  }
>
  <div />
</Intro>
`;
